launcher/logs/LogParser.cpp

```diff
@@ -0,0 +1,358 @@
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, version 3.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "LogParser.h"
+
+#include <QRegularExpression>
+#include "MessageLevel.h"
+
+using namespace Qt::Literals::StringLiterals;
+
+void LogParser::appendLine(QAnyStringView data)
+{
+    if (!m_partialData.isEmpty()) {
+        m_buffer = QString(m_partialData);
+        m_buffer.append("\n");
+        m_partialData.clear();
+    }
+    m_buffer.append(data.toString());
+}
+
+std::optional<LogParser::Error> LogParser::getError()
+{
+    return m_error;
+}
+
+std::optional<LogParser::LogEntry> LogParser::parseAttributes()
+{
+    LogParser::LogEntry entry{
+        "",
+        MessageLevel::Info,
+    };
+    auto attributes = m_parser.attributes();
+
+    for (const auto& attr : attributes) {
+        auto name = attr.name();
+        auto value = attr.value();
+        if (name == "logger"_L1) {
+            entry.logger = value.trimmed().toString();
+        } else if (name == "timestamp"_L1) {
+            if (value.trimmed().isEmpty()) {
+                m_parser.raiseError("log4j:Event Missing required attribute: timestamp");
+                return {};
+            }
+            entry.timestamp = QDateTime::fromSecsSinceEpoch(value.trimmed().toLongLong());
+        } else if (name == "level"_L1) {
+            entry.levelText = value.trimmed().toString();
+            entry.level = MessageLevel::getLevel(entry.levelText);
+        } else if (name == "thread"_L1) {
+            entry.thread = value.trimmed().toString();
+        }
+    }
+    if (entry.logger.isEmpty()) {
+        m_parser.raiseError("log4j:Event Missing required attribute: logger");
+        return {};
+    }
+
+    return entry;
+}
+
+void LogParser::setError()
+{
+    m_error = {
+        m_parser.errorString(),
+        m_parser.error(),
+    };
+}
+
+void LogParser::clearError()
+{
+    m_error = {};  // clear previous error
+}
+
+bool isPotentialLog4JStart(QStringView buffer)
+{
+    static QString target = QStringLiteral("<log4j:event");
+    if (buffer.isEmpty() || buffer[0] != '<') {
+        return false;
+    }
+    auto bufLower = buffer.toString().toLower();
+    return target.startsWith(bufLower) || bufLower.startsWith(target);
+}
+
+std::optional<LogParser::ParsedItem> LogParser::parseNext()
+{
+    clearError();
+
+    if (m_buffer.isEmpty()) {
+        return {};
+    }
+
+    if (m_buffer.trimmed().isEmpty()) {
+        m_buffer.clear();
+        return {};
+    }
```

Comment:
This looks like it would remove empty lines - this was intentionally changed in https://github.com/PrismLauncher/PrismLauncher/pull/1508